applications/RANSApplication/custom_elements/rans_fractional_step_element.cpp

```diff
@@ -0,0 +1,222 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Jordi Cotela
+//
+//  Extended by :    Suneth Warnakulasuriya
+//
+
+// Include base h
+#include "rans_fractional_step_element.h"
+
+namespace Kratos
+{
+template <unsigned int TDim>
+void RansFractionalStepElement<TDim>::CalculateLocalFractionalVelocitySystem(
+    Matrix& rLeftHandSideMatrix,
+    Vector& rRightHandSideVector,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    const auto& r_geometry = this->GetGeometry();
+    const SizeType number_of_nodes = r_geometry.PointsNumber();
+    const SizeType local_size = TDim * number_of_nodes;
+
+    // Check sizes and initialize
+    if (rLeftHandSideMatrix.size1() != local_size) {
+        rLeftHandSideMatrix.resize(local_size, local_size);
+    }
+
+    rLeftHandSideMatrix = ZeroMatrix(local_size, local_size);
+
+    if (rRightHandSideVector.size() != local_size) {
+        rRightHandSideVector.resize(local_size);
+    }
+
+    rRightHandSideVector = ZeroVector(local_size);
+
+    // Shape functions and integration points
+    ShapeFunctionDerivativesArrayType shape_function_derivatives;
+    Matrix shape_functions;
+    Vector gauss_weights;
+    this->CalculateGeometryData(shape_function_derivatives, shape_functions, gauss_weights);
+    const unsigned int number_of_gauss_points = gauss_weights.size();
+
+    Matrix mass_matrix = ZeroMatrix(local_size, local_size);
+
+    const double eta = rCurrentProcessInfo[PRESSURE_COEFFICIENT];
+
+    // Stabilization parameters
+    const double element_size = this->ElementSize();
+    double tau_one, tau_two, density, mass_projection, old_pressure;
+
+    array_1d<double, 3> body_force, momentum_projection, convective_velocity;
+
+    Vector convection_operator(number_of_nodes);
+
+    // Loop on integration points
+    for (unsigned int g = 0; g < number_of_gauss_points; g++) {
+        const double weight = gauss_weights[g];
+        const Vector& r_shape_functions = row(shape_functions, g);
+        const Matrix& r_shape_function_derivatives = shape_function_derivatives[g];
+
+        this->EvaluateInPoint(density, DENSITY, r_shape_functions);
+        this->EvaluateInPoint(mass_projection, DIVPROJ, r_shape_functions);
+        this->EvaluateInPoint(body_force, BODY_FORCE, r_shape_functions);
+        this->EvaluateInPoint(momentum_projection, CONV_PROJ, r_shape_functions);
+        this->EvaluateInPoint(old_pressure, PRESSURE, r_shape_functions, 0);
```

Comment:
Just noting that each one of these calls accesses the nodal database for each one of the Gauss point calculations. Would be nice to introduce a data structure to the FS elements to avoid this performance overhead.